{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample_ID',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Actinomycetales;Actinomycetaceae',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Bifidobacteriales;Bifidobacteriaceae',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Corynebacteriales;Corynebacteriaceae',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Micrococcales;Microbacteriaceae',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Micrococcales;Micrococcaceae',\n",
      "       'Bacteria;Actinobacteriota;Actinobacteria;Propionibacteriales;Propionibacteriaceae',\n",
      "       'Bacteria;Actinobacteriota;Coriobacteriia;Coriobacteriales;Atopobiaceae',\n",
      "       'Bacteria;Actinobacteriota;Coriobacteriia;Coriobacteriales;Coriobacteriaceae',\n",
      "       'Bacteria;Actinobacteriota;Coriobacteriia;Coriobacteriales;Coriobacteriales Incertae Sedis',\n",
      "       'Bacteria;Actinobacteriota;Coriobacteriia;Coriobacteriales;Eggerthellaceae',\n",
      "       'Bacteria;Actinobacteriota;Coriobacteriia;Coriobacteriales;uncultured',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Bacteroidaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Barnesiellaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Marinifilaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Muribaculaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Porphyromonadaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Prevotellaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Rikenellaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;Tannerellaceae',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Bacteroidales;uncultured',\n",
      "       'Bacteria;Bacteroidota;Bacteroidia;Flavobacteriales;Flavobacteriaceae',\n",
      "       'Bacteria;Campilobacterota;Campylobacteria;Campylobacterales;Campylobacteraceae',\n",
      "       'Bacteria;Cyanobacteria;Vampirivibrionia;Gastranaerophilales;Acinetobacter sp. CAG:196',\n",
      "       'Bacteria;Cyanobacteria;Vampirivibrionia;Gastranaerophilales;Heliobacterium sp. N15.MGS-306',\n",
      "       'Bacteria;Cyanobacteria;Vampirivibrionia;Gastranaerophilales;gut metagenome',\n",
      "       'Bacteria;Cyanobacteria;Vampirivibrionia;Gastranaerophilales;uncultured bacterium',\n",
      "       'Bacteria;Cyanobacteria;Vampirivibrionia;Gastranaerophilales;uncultured organism',\n",
      "       'Bacteria;Desulfobacterota;Desulfovibrionia;Desulfovibrionales;Desulfovibrionaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Bacillales;Bacillaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Erysipelotrichales;Erysipelatoclostridiaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Erysipelotrichales;Erysipelotrichaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Izemoplasmatales;uncultured bacterium',\n",
      "       'Bacteria;Firmicutes;Bacilli;Lactobacillales;Carnobacteriaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Lactobacillales;Enterococcaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Lactobacillales;Lactobacillaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Lactobacillales;Leuconostocaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Lactobacillales;Streptococcaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;RF39;gut metagenome',\n",
      "       'Bacteria;Firmicutes;Bacilli;RF39;uncultured bacterium',\n",
      "       'Bacteria;Firmicutes;Bacilli;Staphylococcales;Gemellaceae',\n",
      "       'Bacteria;Firmicutes;Bacilli;Staphylococcales;Staphylococcaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Christensenellales;Christensenellaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia UCG-014;gut metagenome',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia UCG-014;uncultured Clostridiales bacterium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia UCG-014;uncultured Ruminococcaceae bacterium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia UCG-014;uncultured bacterium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia UCG-014;uncultured organism',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia vadinBB60 group;gut metagenome',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia vadinBB60 group;uncultured Thermoanaerobacterales bacterium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia vadinBB60 group;uncultured bacterium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridia vadinBB60 group;uncultured organism',\n",
      "       'Bacteria;Firmicutes;Clostridia;Clostridiales;Clostridiaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Eubacteriales;Anaerofustaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Hungateiclostridiaceae;Ruminiclostridium',\n",
      "       'Bacteria;Firmicutes;Clostridia;Lachnospirales;Defluviitaleaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Lachnospirales;Lachnospiraceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Monoglobales;Monoglobaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;Butyricicoccaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;Ethanoligenenaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;Oscillospiraceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;Ruminococcaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;UCG-010',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;UCG-011',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;[Clostridium] methylpentosum group',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;[Eubacterium] coprostanoligenes group',\n",
      "       'Bacteria;Firmicutes;Clostridia;Oscillospirales;uncultured',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptococcales;Peptococcaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Anaerococcus',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Anaerovoracaceae',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Ezakiella',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Fenollaria',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Peptoniphilus',\n",
      "       'Bacteria;Firmicutes;Clostridia;Peptostreptococcales-Tissierellales;Peptostreptococcaceae',\n",
      "       'Bacteria;Firmicutes;Incertae Sedis;DTU014;uncultured organism',\n",
      "       'Bacteria;Firmicutes;Limnochordia;Hydrogenispora;uncultured Firmicutes bacterium',\n",
      "       'Bacteria;Firmicutes;Negativicutes;Acidaminococcales;Acidaminococcaceae',\n",
      "       'Bacteria;Firmicutes;Negativicutes;Veillonellales-Selenomonadales;Selenomonadaceae',\n",
      "       'Bacteria;Firmicutes;Negativicutes;Veillonellales-Selenomonadales;Veillonellaceae',\n",
      "       'Bacteria;Fusobacteriota;Fusobacteriia;Fusobacteriales;Fusobacteriaceae',\n",
      "       'Bacteria;Patescibacteria;Saccharimonadia;Saccharimonadales;Saccharimonadaceae',\n",
      "       'Bacteria;Proteobacteria;Alphaproteobacteria;Rhodospirillales;uncultured',\n",
      "       'Bacteria;Proteobacteria;Alphaproteobacteria;Rickettsiales;Mitochondria',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Burkholderiales;Comamonadaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Burkholderiales;Neisseriaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Burkholderiales;Oxalobacteraceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Burkholderiales;Sutterellaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Chromatiales;Chromatiaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Hafniaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Pasteurellales;Pasteurellaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Pseudomonadales;Moraxellaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Pseudomonadales;Pseudomonadaceae',\n",
      "       'Bacteria;Proteobacteria;Gammaproteobacteria;Xanthomonadales;Xanthomonadaceae',\n",
      "       'Bacteria;Verrucomicrobiota;Verrucomicrobiae;Verrucomicrobiales;Akkermansiaceae',\n",
      "       'label'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "feature_names must be string, and may not contain [, ] or <",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mValueError\u001B[0m                                Traceback (most recent call last)",
      "Cell \u001B[1;32mIn [18], line 31\u001B[0m\n\u001B[0;32m     28\u001B[0m \u001B[38;5;66;03m# Convert the dataframe to a DMatrix\u001B[39;00m\n\u001B[0;32m     29\u001B[0m x_train, x_test, y_train, y_test \u001B[38;5;241m=\u001B[39m train_test_split(X, Y, test_size\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m0.35\u001B[39m, random_state\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m0\u001B[39m)\n\u001B[1;32m---> 31\u001B[0m dtrain \u001B[38;5;241m=\u001B[39m \u001B[43mxgb\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mDMatrix\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdata\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mx_train\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mlabel\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43my_train\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m     32\u001B[0m dtest \u001B[38;5;241m=\u001B[39m xgb\u001B[38;5;241m.\u001B[39mDMatrix(data\u001B[38;5;241m=\u001B[39mx_test, label\u001B[38;5;241m=\u001B[39my_test)\n\u001B[0;32m     34\u001B[0m \u001B[38;5;66;03m#dtrain = xgb.DMatrix(data=X_train, label=Y_train)\u001B[39;00m\n\u001B[0;32m     35\u001B[0m \u001B[38;5;66;03m#dtest = xgb.DMatrix(data=X_test, label=Y_test)\u001B[39;00m\n\u001B[0;32m     36\u001B[0m \n\u001B[0;32m     37\u001B[0m \u001B[38;5;66;03m# Create the parameter dictionary for the XGBoost model\u001B[39;00m\n\u001B[0;32m     38\u001B[0m \u001B[38;5;66;03m#params = {'objective': 'reg:squarederror', 'eval_metric': 'rmse'}\u001B[39;00m\n",
      "File \u001B[1;32mc:\\users\\luca\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\xgboost\\core.py:620\u001B[0m, in \u001B[0;36mrequire_keyword_args.<locals>.throw_if.<locals>.inner_f\u001B[1;34m(*args, **kwargs)\u001B[0m\n\u001B[0;32m    618\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m k, arg \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mzip\u001B[39m(sig\u001B[38;5;241m.\u001B[39mparameters, args):\n\u001B[0;32m    619\u001B[0m     kwargs[k] \u001B[38;5;241m=\u001B[39m arg\n\u001B[1;32m--> 620\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m func(\u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mkwargs)\n",
      "File \u001B[1;32mc:\\users\\luca\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\xgboost\\core.py:766\u001B[0m, in \u001B[0;36mDMatrix.__init__\u001B[1;34m(self, data, label, weight, base_margin, missing, silent, feature_names, feature_types, nthread, group, qid, label_lower_bound, label_upper_bound, feature_weights, enable_categorical)\u001B[0m\n\u001B[0;32m    754\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mset_info(\n\u001B[0;32m    755\u001B[0m     label\u001B[38;5;241m=\u001B[39mlabel,\n\u001B[0;32m    756\u001B[0m     weight\u001B[38;5;241m=\u001B[39mweight,\n\u001B[1;32m   (...)\u001B[0m\n\u001B[0;32m    762\u001B[0m     feature_weights\u001B[38;5;241m=\u001B[39mfeature_weights,\n\u001B[0;32m    763\u001B[0m )\n\u001B[0;32m    765\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m feature_names \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m--> 766\u001B[0m     \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfeature_names\u001B[49m \u001B[38;5;241m=\u001B[39m feature_names\n\u001B[0;32m    767\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m feature_types \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[0;32m    768\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mfeature_types \u001B[38;5;241m=\u001B[39m feature_types\n",
      "File \u001B[1;32mc:\\users\\luca\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\xgboost\\core.py:1164\u001B[0m, in \u001B[0;36mDMatrix.feature_names\u001B[1;34m(self, feature_names)\u001B[0m\n\u001B[0;32m   1160\u001B[0m \u001B[38;5;66;03m# prohibit to use symbols may affect to parse. e.g. []<\u001B[39;00m\n\u001B[0;32m   1161\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28mall\u001B[39m(\u001B[38;5;28misinstance\u001B[39m(f, \u001B[38;5;28mstr\u001B[39m) \u001B[38;5;129;01mand\u001B[39;00m\n\u001B[0;32m   1162\u001B[0m            \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28many\u001B[39m(x \u001B[38;5;129;01min\u001B[39;00m f \u001B[38;5;28;01mfor\u001B[39;00m x \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mset\u001B[39m((\u001B[38;5;124m'\u001B[39m\u001B[38;5;124m[\u001B[39m\u001B[38;5;124m'\u001B[39m, \u001B[38;5;124m'\u001B[39m\u001B[38;5;124m]\u001B[39m\u001B[38;5;124m'\u001B[39m, \u001B[38;5;124m'\u001B[39m\u001B[38;5;124m<\u001B[39m\u001B[38;5;124m'\u001B[39m)))\n\u001B[0;32m   1163\u001B[0m            \u001B[38;5;28;01mfor\u001B[39;00m f \u001B[38;5;129;01min\u001B[39;00m feature_names):\n\u001B[1;32m-> 1164\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mfeature_names must be string, and may not contain [, ] or <\u001B[39m\u001B[38;5;124m'\u001B[39m)\n\u001B[0;32m   1165\u001B[0m feature_names_bytes \u001B[38;5;241m=\u001B[39m [\u001B[38;5;28mbytes\u001B[39m(f, encoding\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mutf-8\u001B[39m\u001B[38;5;124m'\u001B[39m) \u001B[38;5;28;01mfor\u001B[39;00m f \u001B[38;5;129;01min\u001B[39;00m feature_names]\n\u001B[0;32m   1166\u001B[0m c_feature_names \u001B[38;5;241m=\u001B[39m (ctypes\u001B[38;5;241m.\u001B[39mc_char_p \u001B[38;5;241m*\u001B[39m\n\u001B[0;32m   1167\u001B[0m                    \u001B[38;5;28mlen\u001B[39m(feature_names_bytes))(\u001B[38;5;241m*\u001B[39mfeature_names_bytes)\n",
      "\u001B[1;31mValueError\u001B[0m: feature_names must be string, and may not contain [, ] or <"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score\n",
    "import xgboost as xgb\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "from sklearn.metrics import f1_score, matthews_corrcoef\n",
    "\n",
    "#df = pd.read_csv('../HackathonMicrobiomeData/CAD/ClassCAD_train.csv')\n",
    "#df = pd.read_csv('../src/SMOTEclass.csv')\n",
    "df = pd.read_csv(\"DataMatrices/order/unionMatrix_TreeBasedClassifier_Normalized_order.csv\")\n",
    "\n",
    "Y = np.array(df[[\"label\"]]).ravel()\n",
    "#X = df.drop([\"label\"], axis = 1)\n",
    "X = df.drop([\"sample_ID\", \"label\"], axis = 1)\n",
    "\n",
    "# X_test = X.iloc[:250]\n",
    "# print(X_test.shape)\n",
    "# Y_test = Y[:250]\n",
    "# print(Y_test.shape)\n",
    "# X_train = X.iloc[251:]\n",
    "# print(X_train.shape)\n",
    "# Y_train = Y[251:]\n",
    "# print(Y_train.shape)\n",
    "\n",
    "print(df.columns)\n",
    "\n",
    "\n",
    "# Convert the dataframe to a DMatrix\n",
    "x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.35, random_state=0)\n",
    "\n",
    "dtrain = xgb.DMatrix(data=x_train, label=y_train)\n",
    "dtest = xgb.DMatrix(data=x_test, label=y_test)\n",
    "\n",
    "#dtrain = xgb.DMatrix(data=X_train, label=Y_train)\n",
    "#dtest = xgb.DMatrix(data=X_test, label=Y_test)\n",
    "\n",
    "# Create the parameter dictionary for the XGBoost model\n",
    "#params = {'objective': 'reg:squarederror', 'eval_metric': 'rmse'}\n",
    "params = {\n",
    "    'objective': 'multi:softmax',\n",
    "    'max_depth': 20,\n",
    "    'learning_rate': 0.1,\n",
    "    'n_estimators': 5,\n",
    "    'seed': 123,\n",
    "    'num_class': 5 # specify the number of classes\n",
    "}\n",
    "\n",
    "# Create the XGBoost model\n",
    "model = xgb.train(params, dtrain)\n",
    "\n",
    "# Make predictions on the test set\n",
    "predictions = model.predict(dtest)\n",
    "\n",
    "# Evaluate the model\n",
    "#accuracy = accuracy_score(y_test, predictions)\n",
    "accuracy = accuracy_score(y_test, predictions)\n",
    "print('Accuracy:', accuracy)\n",
    "\n",
    "print(\"MCC\", matthews_corrcoef(y_test,predictions))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
